esphome/components/voice_kit/voice_kit.cpp

```diff
@@ -21,18 +25,188 @@ void VoiceKit::setup() {
     if (!this->dfu_get_version_()) {
       ESP_LOGE(TAG, "Communication with Voice Kit failed");
       this->mark_failed();
+    } else if (!versions_match() && this->firmware_bin_ != nullptr && this->firmware_bin_length_) {
+      ESP_LOGW(TAG, "XMOS firmware version is incorrect -- updating...");
+      this->flash();
     }
   });
 }
 
 void VoiceKit::dump_config() {
   ESP_LOGCONFIG(TAG, "Voice Kit:");
-  LOG_PIN("  Reset Pin: ", this->reset_pin_);
   LOG_I2C_DEVICE(this);
+  LOG_PIN("  Reset Pin: ", this->reset_pin_);
+  if (this->firmware_version_major_ || this->firmware_version_minor_ || this->firmware_version_patch_) {
+    ESP_LOGCONFIG(TAG, "  XMOS firmware version: %u.%u.%u", this->firmware_version_major_,
+                  this->firmware_version_minor_, this->firmware_version_patch_);
+  }
+}
+
+void VoiceKit::flash() {
+  if (this->firmware_bin_ == nullptr || !this->firmware_bin_length_) {
+    ESP_LOGE(TAG, "Firmware invalid");
+    return;
+  }
+
+  ESP_LOGI(TAG, "Starting update...");
+  // #ifdef USE_OTA_STATE_CALLBACK
+  //   this->state_callback_.call(ota::OTA_STARTED, 0.0f, 0);
+  // #endif
+
+  auto ota_status = this->do_ota_();
+
+  //   switch (ota_status) {
```

Comment:
:face_with_peeking_eye:

Reply:
Haha, I left this in because I knew you'd comment on...the comments... 😅

We should talk more about if we want to tie in some triggers or something to provide some indication that this process is happening -- it hangs the device for about 12 seconds while the update is in progress.

I also wasn't sure if doing the update as a part of `setup()` is a good idea or not. In my testing, it doesn't _seem_ to cause any problems, but I was hoping to chat a bit about if this approach is reasonable or if it might be better to defer it to some point later on.

Reply:
If its gonna take that long, I would start in `setup`, but do parts in a `loop` and also implement `can_proceed()` and return false until the flashing is done. 
This would mean the entire `setup` process of ESPHome would be "halted" until the flashing was done, but any `loop` functions of previously set up components will still run correctly.

Reply:
I'll experiment more with that, but some earlier testing I was doing suggests that the process is quite time-sensitive. If we wait too long to send the next block, the DFU process will time out and fail on the XMOS side.

Reply:
Got it working in [`4644c36` (#68)](https://github.com/esphome/voice-kit/pull/68/commits/4644c363ba9cc5a4eda588acbc42fa931de9522b) 🎉